### ISSUE TITLE
AP Calculation

### ISSUE THREAD

**Issue by ksv87** [ultralytics/ultralytics#17270]
### Discussed in https://github.com/orgs/ultralytics/discussions/17269

<div type='discussions-op-text'>

<sup>Originally posted by **ksv87** October 30, 2024</sup>
Why is AP considered this way? The P-R curve calculation doesn't look right
https://github.com/ultralytics/ultralytics/blob/b8c90baffee06b7b162cb29bd94383a693a42744/ultralytics/utils/metrics.py#L514-L516

![image](https://github.com/user-attachments/assets/ff2d46a3-a56b-4e2e-96f8-afd83a07cae7)

Correctly it should be like this
```
    # Append sentinel values to beginning and end
    mrec = np.concatenate(([0.0], recall, [recall[-1]], [1.0]))
    mpre = np.concatenate(([1.0], precision, [0.0], [0.0]))
```

![image](https://github.com/user-attachments/assets/074b0c15-c245-4dff-9665-0b0ab13a2322)

Calculation example https://pyimagesearch.com/2022/05/02/mean-average-precision-map-using-the-coco-evaluator/
![image](https://github.com/user-attachments/assets/9fb768f0-a511-451d-9df6-6e752a617635)
</div>

**Comment by ksv87** [comment]
What changes besides the fact that instead of 11 points it becomes 101?
The problem is that the AP via COCO Evaluator is considered less than that of Ultralytics

If you look at the first 2 graphs - this is the same model, just replaced the code that I specified.
After the correction, the AP became more similar to the values obtained from COCO Evaluator

The area under the curve should be considered so that only real values obtained from the network are taken into account, isn't it? In my example, the network didn't actually return any values with Recall above 0.65. Why the area under the curve in Ultralytics captures Recall values greater than 0.65?

Also, the AP should never exceed either Recall or Precision, shouldn't it?

**Comment by Y-T-G** [comment]
I get this with your modification on COCO dataset.

```
In [2]: !yolo val model=yolo11n.pt  data=coco.yaml
Ultralytics 8.3.24 🚀 Python-3.10.14 torch-2.4.1+cu121 CUDA:0 (Tesla T4, 14916MiB)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgro
                 Class     Images  Instances      Box(P          R      mAP5
                   all       5000      36335      0.653      0.504      0.546      0.388
                person       2693      10777       0.76      0.676       0.75      0.517
               bicycle        149        314      0.692      0.404      0.475      0.272
                   car        535       1918      0.658       0.53       0.58      0.371
            motorcycle        159        367      0.765      0.599      0.672      0.439
              airplane         97        143      0.803      0.797      0.869      0.684
                   bus        189        283      0.772      0.707      0.771      0.646
                 train        157        190      0.827      0.811      0.862      0.678
                 truck        250        414      0.585      0.415      0.476       0.32
                  boat        121        424      0.583      0.354      0.411      0.217
         traffic light        191        634      0.647      0.345      0.411      0.213
          fire hydrant         86        101      0.867      0.743      0.803      0.633
             stop sign         69         75      0.712      0.613      0.659      0.595
         parking meter         37         60      0.698       0.55       0.61      0.463
                 bench        235        411      0.575       0.28      0.323      0.214
                  bird        125        427      0.618      0.382      0.436      0.276
                   cat        184        202      0.782      0.861      0.881      0.697
                   dog        177        218      0.761      0.743      0.777      0.645
                 horse        128        272      0.764      0.706      0.769      0.588
                 sheep         65        354      0.703      0.692      0.743      0.503
                   cow         87        372      0.719      0.679       0.73      0.515
              elephant         89        252      0.729      0.837      0.841       0.65
                  bear         49         71      0.851      0.859      0.909       0.74
                 zebra         85        266      0.804      0.838      0.892       0.67
               giraffe        101        232      0.859      0.845       0.89      0.696
              backpack        228        371      0.491      0.197        0.2      0.103
              umbrella        174        407      0.652      0.537      0.577      0.385
               handbag        292        540      0.476      0.128      0.163     0.0861
                   tie        145        252      0.629      0.413      0.445      0.275
              suitcase        105        299      0.636      0.472      0.533       0.37
               frisbee         84        115      0.787      0.783      0.813      0.608
                  skis        120        241      0.578      0.336      0.393      0.209
             snowboard         49         69      0.607      0.391      0.402      0.275
           sports ball        169        260      0.692      0.488      0.504       0.36
                  kite         91        327       0.64      0.541      0.595      0.405
          baseball bat         97        145      0.571      0.393      0.414      0.229
        baseball glove        100        148      0.682        0.5      0.535      0.318
            skateboard        127        179      0.706      0.648       0.67      0.464
             surfboard        149        267      0.645      0.498      0.536      0.338
         tennis racket        167        225      0.747      0.671      0.695      0.414
                bottle        379       1013      0.622      0.414      0.467      0.307
            wine glass        110        341      0.706      0.334      0.422      0.274
                   cup        390        895      0.583      0.449      0.499      0.358
                  fork        155        215      0.588      0.339      0.383      0.266
                 knife        181        325      0.471      0.154      0.201       0.12
                 spoon        153        253      0.398      0.154      0.159     0.0969
                  bowl        314        623      0.608      0.496      0.525      0.387
                banana        103        370      0.499      0.297      0.359      0.221
                 apple         76        236      0.399      0.237      0.231      0.163
              sandwich         98        177       0.56      0.503      0.481      0.358
                orange         85        285      0.473      0.399      0.384      0.296
              broccoli         71        312      0.498      0.356      0.384      0.211
                carrot         81        365      0.438      0.323      0.286      0.177
               hot dog         51        125      0.649      0.464       0.49      0.352
                 pizza        153        284      0.703      0.641      0.688      0.519
                 donut         62        328      0.611      0.507      0.545      0.422
                  cake        124        310      0.601      0.439      0.517      0.353
                 chair        580       1771      0.591      0.362      0.416      0.261
                 couch        195        261      0.623      0.582      0.631      0.473
          potted plant        172        342      0.549      0.395      0.392      0.224
                   bed        149        163      0.627      0.619       0.66       0.48
          dining table        501        695      0.545      0.449      0.446      0.308
                toilet        149        179      0.732      0.763       0.82      0.664
                    tv        207        288      0.721       0.67      0.722      0.543
                laptop        183        231      0.668       0.68       0.71      0.586
                 mouse         88        106      0.716      0.717       0.74      0.538
                remote        145        283      0.507      0.279      0.321      0.185
              keyboard        106        153      0.636      0.582       0.64      0.479
            cell phone        214        262      0.588      0.431      0.438      0.301
             microwave         54         55      0.681        0.6      0.669      0.517
                  oven        115        143      0.637       0.49      0.554       0.38
               toaster          8          9          1      0.427       0.69      0.454
                  sink        187        225      0.662      0.511      0.539      0.363
          refrigerator        101        126      0.769      0.643      0.722       0.57
                  book        230       1129        0.5      0.118      0.214      0.105
                 clock        204        267      0.755      0.625      0.654      0.455
                  vase        137        274      0.589      0.478      0.476      0.324
              scissors         28         36      0.529      0.306      0.325      0.274
            teddy bear         94        190      0.688      0.574      0.625      0.454
            hair drier          9         11          1          0   0.000211   4.22e-05
            toothbrush         34         57       0.47      0.281       0.25      0.162
Speed: 0.1ms preprocess, 2.8ms inference, 0.0ms loss, 0.6ms postprocess per image
Saving /ultralytics/runs/detect/val3/predictions.json...

Evaluating pycocotools mAP using /ultralytics/runs/detect/val3/predictions.json and /datasets/coco/annotations/instances_val2017.json...
loading annotations into memory...
Done (t=0.84s)
creating index...
index created!
Loading and preparing results...
DONE (t=4.60s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=50.72s).
Accumulating evaluation results...
DONE (t=11.61s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.394
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.553
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.428
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.199
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.432
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.540
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.370
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.781
```

and this is with the original:

```
In [4]: !yolo val model=yolo11n.pt  data=coco.yaml
Ultralytics 8.3.24 🚀 Python-3.10.14 torch-2.4.1+cu121 CUDA:0 (Tesla T4, 14916MiB)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100%|████████
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████|
                   all       5000      36335      0.653      0.504      0.549      0.392
                person       2693      10777       0.76      0.676      0.753      0.523
               bicycle        149        314      0.692      0.404      0.478      0.275
                   car        535       1918      0.658       0.53      0.583      0.375
            motorcycle        159        367      0.765      0.599      0.674      0.443
              airplane         97        143      0.803      0.797      0.869      0.686
                   bus        189        283      0.772      0.707      0.773       0.65
                 train        157        190      0.827      0.811      0.863      0.681
                 truck        250        414      0.585      0.415      0.477      0.322
                  boat        121        424      0.583      0.354      0.414       0.22
         traffic light        191        634      0.647      0.345      0.415      0.217
          fire hydrant         86        101      0.867      0.743      0.806      0.638
             stop sign         69         75      0.712      0.613      0.662      0.599
         parking meter         37         60      0.698       0.55      0.613      0.467
                 bench        235        411      0.575       0.28      0.325      0.216
                  bird        125        427      0.618      0.382      0.441       0.28
                   cat        184        202      0.782      0.861      0.882      0.702
                   dog        177        218      0.761      0.743      0.779      0.649
                 horse        128        272      0.764      0.706       0.77      0.592
                 sheep         65        354      0.703      0.692      0.745      0.507
                   cow         87        372      0.719      0.679      0.733       0.52
              elephant         89        252      0.729      0.837      0.842      0.655
                  bear         49         71      0.851      0.859      0.911      0.745
                 zebra         85        266      0.804      0.838      0.893      0.674
               giraffe        101        232      0.859      0.845      0.891        0.7
              backpack        228        371      0.491      0.197      0.202      0.105
              umbrella        174        407      0.652      0.537      0.581      0.391
               handbag        292        540      0.476      0.128      0.165     0.0883
                   tie        145        252      0.629      0.413      0.449      0.279
              suitcase        105        299      0.636      0.472      0.536      0.374
               frisbee         84        115      0.787      0.783      0.814      0.611
                  skis        120        241      0.578      0.336      0.396      0.212
             snowboard         49         69      0.607      0.391      0.404      0.277
           sports ball        169        260      0.692      0.488      0.511      0.367
                  kite         91        327       0.64      0.541      0.599      0.411
          baseball bat         97        145      0.571      0.393      0.417      0.232
        baseball glove        100        148      0.682        0.5      0.539      0.321
            skateboard        127        179      0.706      0.648      0.672      0.467
             surfboard        149        267      0.645      0.498       0.54      0.342
         tennis racket        167        225      0.747      0.671      0.697      0.417
                bottle        379       1013      0.622      0.414      0.471      0.311
            wine glass        110        341      0.706      0.334      0.427      0.279
                   cup        390        895      0.583      0.449      0.502      0.362
                  fork        155        215      0.588      0.339      0.385      0.268
                 knife        181        325      0.471      0.154      0.204      0.123
                 spoon        153        253      0.398      0.154      0.162     0.0993
                  bowl        314        623      0.608      0.496      0.527      0.391
                banana        103        370      0.499      0.297      0.362      0.225
                 apple         76        236      0.399      0.237      0.235      0.167
              sandwich         98        177       0.56      0.503      0.483      0.361
                orange         85        285      0.473      0.399      0.389      0.302
              broccoli         71        312      0.498      0.356      0.385      0.213
                carrot         81        365      0.438      0.323      0.289       0.18
               hot dog         51        125      0.649      0.464      0.494      0.356
                 pizza        153        284      0.703      0.641       0.69      0.523
                 donut         62        328      0.611      0.507      0.549      0.427
                  cake        124        310      0.601      0.439       0.52      0.357
                 chair        580       1771      0.591      0.362      0.419      0.265
                 couch        195        261      0.623      0.582      0.632      0.476
          potted plant        172        342      0.549      0.395      0.395      0.227
                   bed        149        163      0.627      0.619      0.661      0.483
          dining table        501        695      0.545      0.449      0.449      0.312
                toilet        149        179      0.732      0.763      0.822      0.668
                    tv        207        288      0.721       0.67      0.725      0.548
                laptop        183        231      0.668       0.68      0.713       0.59
                 mouse         88        106      0.716      0.717      0.742      0.542
                remote        145        283      0.507      0.279      0.323      0.187
              keyboard        106        153      0.636      0.582      0.641      0.481
            cell phone        214        262      0.588      0.431       0.44      0.304
             microwave         54         55      0.681        0.6       0.67       0.52
                  oven        115        143      0.637       0.49      0.556      0.383
               toaster          8          9          1      0.427      0.691      0.455
                  sink        187        225      0.662      0.511      0.541      0.367
          refrigerator        101        126      0.769      0.643      0.724      0.573
                  book        230       1129        0.5      0.118      0.217      0.108
                 clock        204        267      0.755      0.625      0.657      0.459
                  vase        137        274      0.589      0.478      0.478      0.328
              scissors         28         36      0.529      0.306      0.327      0.276
            teddy bear         94        190      0.688      0.574      0.627      0.458
            hair drier          9         11          1          0    0.00103   0.000207
            toothbrush         34         57       0.47      0.281      0.252      0.164
Speed: 0.1ms preprocess, 2.8ms inference, 0.0ms loss, 0.6ms postprocess per image
Saving /ultralytics/runs/detect/val4/predictions.json...

Evaluating pycocotools mAP using /ultralytics/runs/detect/val4/predictions.json and /datasets/coco/annotations/instances_val2017.json...
loading annotations into memory...
Done (t=0.85s)
creating index...
index created!
Loading and preparing results...
DONE (t=4.71s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=51.00s).
Accumulating evaluation results...
DONE (t=9.76s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.394
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.553
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.428
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.199
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.432
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.540
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.370
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.781
```

The original is closer to the one from COCOAPI.

**Comment by ksv87** [comment]
@Y-T-G  Thanks for this test

As I understand it, the more Recall and Precision - the smaller the difference with and without correction, this is relevant for datasets and models on which the Recall is low, because if the Recall is high, the curve reaches almost the end and the area does not change much

**Comment by Y-T-G** [comment]
I don't see the precision getting forced to 0 after highest recall value in the COCO code either.

https://github.com/cocodataset/cocoapi/blob/8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9/PythonAPI/pycocotools/cocoeval.py#L315

Maybe @MiXaiLL76 knows better since they worked on [faster_coco_eval](https://github.com/MiXaiLL76/faster_coco_eval).

**Comment by ksv87** [comment]
In my test  i run `yolo val model=runs\detect\yolov5n6_1280\weights\best.pt data="D:\projects\test\dataset\data.yaml" conf=0.25 imgsz=1280 save_json`

YOLOv8 original code 
```
Ultralytics 8.3.25 🚀 Python-3.12.3 torch-2.5.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4070 SUPER, 12282MiB)
YOLOv5n6u summary (fused): 253 layers, 4,127,096 parameters, 0 gradients, 7.2 GFLOPs
val: Scanning D:\projects\test\dataset\val2017.cache... 886 images, 9 backgrounds, 0 corrupt: 100%|██████████| 895/895 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 56/56 [00:13<00:00,  4.23it/s]
                   all        895      11047      0.583      0.526      0.563      0.383
                class1        280       1599      0.515      0.397      0.469      0.322
                class2        220        776      0.334       0.33      0.302      0.193
                class3        249       1279      0.678        0.7      0.732      0.548
                class4        106       3654      0.512      0.332      0.435       0.19
                class5        182        617       0.55      0.602      0.559      0.311
                class6        665       3122       0.91      0.797      0.885      0.738
Speed: 0.9ms preprocess, 3.8ms inference, 0.0ms loss, 1.0ms postprocess per image
Saving runs\detect\val33\predictions.json...
Results saved to runs\detect\val33
```

next i try COCO Eval
```
from pycocotools.cocoeval import COCOeval

cocoAnnotation = COCO(annotation_file="D:/projects/test/dataset/Annotations/instances_val2017.json")
cocovalPrediction = cocoAnnotation.loadRes("runs/detect/val33/predictions.json")

cocoEval = COCOeval(cocoAnnotation, cocovalPrediction, "bbox")
cocoEval.evaluate()
cocoEval.accumulate()
cocoEval.summarize()
```

map strong different

```
DONE (t=0.15s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.288
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.443
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.298
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.105
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.315
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.416
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.104
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.314
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.157
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.406
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.507
```

with my modification `yolo val model=runs\detect\yolov5n6_1280\weights\best.pt data="D:\projects\test\dataset\data.yaml" conf=0.25 imgsz=1280 save_json`

```
Ultralytics 8.3.25 🚀 Python-3.12.3 torch-2.5.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4070 SUPER, 12282MiB)
YOLOv5n6u summary (fused): 253 layers, 4,127,096 parameters, 0 gradients, 7.2 GFLOPs
val: Scanning D:\projects\test\dataset\val2017.cache... 886 images, 9 backgrounds, 0 corrupt: 100%|██████████| 895/895 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 56/56 [00:13<00:00,  4.11it/s]
                   all        895      11047      0.583      0.526      0.443      0.286
                class1        280       1599      0.515      0.397      0.318      0.205
                class2        220        776      0.334       0.33      0.192      0.114
                class3        249       1279      0.678        0.7      0.634      0.444
                class4        106       3654      0.512      0.332       0.27      0.104
                class5        182        617       0.55      0.602      0.451      0.225
                class6        665       3122       0.91      0.797       0.79      0.626
Speed: 0.6ms preprocess, 4.1ms inference, 0.0ms loss, 1.0ms postprocess per image
Saving runs\detect\val34\predictions.json...
Results saved to runs\detect\val34
```

this is closer to COCO Eval

> I don't see the precision getting forced to 0 after highest recall value in the COCO code either.

COCO Eval not use np.interp, is use np.searchsorted
https://github.com/cocodataset/cocoapi/blob/8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9/PythonAPI/pycocotools/cocoeval.py#L402-L410

it describe her https://github.com/ultralytics/ultralytics/issues/14063

**Comment by MiXaiLL76** [comment]
@ksv87 Can you validate your dataset and present it in the format:
https://nbviewer.org/github/MiXaiLL76/faster_coco_eval/blob/main/examples/curve_example.ipynb

Let's see how my library builds pre-rec curve

**Comment by ksv87** [comment]
@MiXaiLL76 i validate with fast_coco_eval
![newplot](https://github.com/user-attachments/assets/dc5eceb0-6eee-4c06-9728-7900ab857d5d)

**Comment by MiXaiLL76** [comment]
Still, I think that for working under Windows you will have to wait for the next release. There is some bug with fp32 now.

**Comment by ksv87** [comment]
with useCats
![newplot1](https://github.com/user-attachments/assets/b7e0c6a7-b5cf-4f30-bd67-82561fb0b213)

**Comment by MiXaiLL76** [comment]
Can you provide:
- instances_val2017.json
- predictions.json

?

**Comment by ksv87** [comment]
[instances_val2017.json](https://github.com/user-attachments/files/17584928/instances_val2017.json)
[predictions.json_fix.json](https://github.com/user-attachments/files/17584929/predictions.json_fix.json)
yes

**Comment by MiXaiLL76** [comment]
> [instances_val2017.json](https://github.com/user-attachments/files/17584928/instances_val2017.json) [predictions.json_fix.json](https://github.com/user-attachments/files/17584929/predictions.json_fix.json) yes

![изображение](https://github.com/user-attachments/assets/246bdf6f-0ce4-4efc-b476-02456b1e1c15)
I get identical metrics as you, on the new version. Good.

My AP\AR calculation function is very different because it is written in C++.
But the graph looks correct in general.
But the first picture you presented in the post header looks wrong.

But this is my personal opinion. Perhaps your correction is in place.
I still don't quite understand why the numbers on my graph are very different from your second graph (after the correction).
Perhaps min_score or threshold_iou affect this.

```py
precision_list = self.eval["precision"][:, :, category_id, :, :].ravel()
recall_list = self.recThrs
scores = self.eval["scores"][:, :, category_id, :, :].ravel()
auc = round(COCOeval_faster.calc_auc(recall_list, precision_list), 4)
```

calc_auc taken from here
https://towardsdatascience.com/how-to-efficiently-implement-area-under-precision-recall-curve-pr-auc-a85872fd7f14

**Comment by ksv87** [comment]
Потому что я забыл какую модель использовал в первом посте, в последнем примере я использовал другую модель

original
![image](https://github.com/user-attachments/assets/d39cc170-28c1-47ce-a583-6d395890c840)

fixed
![image](https://github.com/user-attachments/assets/f420ac94-f915-4cd7-8d87-c4061990da03)

**Comment by glenn-jocher** [comment]
Thank you for sharing your results. It seems like the differences in AP calculations are due to the variations in the models and evaluation methods used. If you have further questions or need more insights, feel free to explore our documentation or reach out to the community for additional support.

**Comment by ksv87** [comment]
> Thank you for sharing your results. It seems like the differences in AP calculations are due to the variations in the models and evaluation methods used. If you have further questions or need more insights, feel free to explore our documentation or reach out to the community for additional support.

Last 3 graphics is 1 model

**Comment by glenn-jocher** [comment]
Thank you for clarifying. The differences might still arise from the evaluation methods or settings. If you need further assistance, please refer to our documentation or community forums for more detailed guidance.

**Comment by ksv87** [comment]
ultralytics framework have one hardcode option:  **method = "interp"  # methods: 'continuous', 'interp'** 


https://github.com/ultralytics/ultralytics/blob/b8c90baffee06b7b162cb29bd94383a693a42744/ultralytics/utils/metrics.py#L522-L528

default is **interp**, mAP different with COCO Eval and Fast COCO Eval
```
 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 56/56 [00:13<00:00,  4.08it/s]
                   all        895      11047      0.583      0.526      0.563      0.383
                class1        280       1599      0.515      0.397      0.469      0.322
                class2        220        776      0.334       0.33      0.302      0.193
                class3        249       1279      0.678        0.7      0.732      0.548
                class4        106       3654      0.512      0.332      0.435       0.19
                class5        182        617       0.55      0.602      0.559      0.311
                class6        665       3122       0.91      0.797      0.885      0.738
```

if change it to **continuous**, mAP more similar to COCO Eval and Fast COCO Eval
```
Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 56/56 [00:14<00:00,  4.00it/s]
                   all        895      11047      0.583      0.526      0.442      0.286
                class1        280       1599      0.515      0.397      0.316      0.205
                class2        220        776      0.334       0.33      0.192      0.114
                class3        249       1279      0.678        0.7      0.631      0.444
                class4        106       3654      0.512      0.332      0.268      0.103
                class5        182        617       0.55      0.602      0.452      0.225
                class6        665       3122       0.91      0.797      0.793      0.627
```

COCO Eval
```
DONE (t=0.15s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.288
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.443
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.298
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.105
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.315
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.416
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.104
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.314
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.157
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.406
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.507
```

This is a big difference, which creates a strong hindrance in comparing ultralitycs models with other models in which metrics are calculated using COCO Eval.

**Comment by Y-T-G** [comment]
@Laughing-q Should it be changed to use `continuous` mode?

**Comment by lushikai** [comment]
@glenn-jocher Excuse you here for a while. Can you help explain what `preprocess`, `inference`, `loss`and`postprocess per image` mean? Which one should i chose for the `val` mode and the metrics for model performance. Due to different models, `postprocess per image` is under 0 and `loss` is 0 as follows. Whether is it correct? Thanks for your time! 😸 
![image](https://github.com/user-attachments/assets/c8968c6e-4f4a-4bd7-878e-fe8d0090b9fe)
![image](https://github.com/user-attachments/assets/63c2db9b-c6bd-4ae9-a514-33907e7aef4e)

**Comment by Skillnoob** [comment]
@lushikai You should have opened a new issue for that, this is offtopic.
The values mean exactly what their name is. Preprocess is the time it took to pre process the image, inference is the time it took for the model to process it and postprocess is the time to get the results processed. Loss is irrelevant, its pretty much always 0.

**Comment by lushikai** [comment]
@Skillnoob Sorry. Thanks for your explanation! 😄

**Comment by MiXaiLL76** [comment]
>  In faster-coco-eval, do we get the final mAP by calculating the areas under each curve?

In my library, auc is calculated separately, "in parallel" with the calculation of mAP. mAP is calculated by the same function, almost without changes, as in pycocotools.

![изображение](https://github.com/user-attachments/assets/5c77f5ca-765e-466f-9bfe-eda1c406fdad)

**Comment by MiXaiLL76** [comment]
mAP-.5-.95 - considered this way 

```py
import numpy as np

cocoGt = COCO(...)
cocoDt = cocoGt.loadRes(...)

cocoEval = COCOeval_faster(cocoGt, cocoDt, ....)
cocoEval.evaluate()
cocoEval.accumulate()

aind = cocoEval.params.areaRngLbl.index("all")
mind = cocoEval.params.maxDets.index(cocoEval.params.maxDets[-1])

precision_list = cocoEval.eval["precision"][:, :, :, aind, mind]
recall_list = cocoEval.params.recThrs

print(precision_list.shape) # iou_count * recall * category

np.mean(precision_list[precision_list > -1]) # AP
```

In my case, in my test dataset it turns out to be

0.6947194719471946

If you try to use the function from ultralytics, you get something like this:

https://github.com/ultralytics/ultralytics/blob/b8c90baffee06b7b162cb29bd94383a693a42744/ultralytics/utils/metrics.py#L501

```py
total_ap = []
for iou_i in range(cocoEval.params.iouThrs.shape[0]):
for category_i in range(len(cocoEval. params.cat_ids)):
pre = precision_list[iou_i, :, category_i]
ap, _, _ = compute_ap(recall_list, pre)
total_ap.append(ap)

np.mean(total_ap)
```

i.e. calculated for all IOUs and categories, then averaged.

0.69208333333333334

Close, but not hit. In calculations pycocotools does not participate recall.

**Comment by glenn-jocher** [comment]
The slight difference in mAP calculations between your method and the Ultralytics function may stem from averaging across all IoUs and categories. This approach ensures a comprehensive evaluation across varying detection thresholds.

**Comment by glenn-jocher** [comment]
The slight difference in mAP calculations could be due to variations in how averages are computed across IoU thresholds and categories. Both methods are valid, but they may handle edge cases or missing values differently.

**Comment by WCUSTC** [comment]
In my task, the AP50 metric calculated by ultralytics far exceeds the metric obtained by mainstream calculation methods

**Comment by Y-T-G** [comment]
> @ksv87 @Y-T-G @MiXaiLL76 Hi guys! after taking a look at this one, actually the correction made by @ksv87 looks good to me!
> 
> ```python
>     # Append sentinel values to beginning and end
>     mrec = np.concatenate(([0.0], recall, [recall[-1]], [1.0]))
>     mpre = np.concatenate(([1.0], precision, [0.0], [0.0]))
> ```
> 
> And it does make the mAP more consistent with pycocotools/faster-coco-eval when using higher confidence score thresholds i.e. 0.1, 0.2. However we need to figure out why the mAP decreases to 38.8 when applying it to coco val set with `con=0.001`. Do you guys have any ideas? From the posts above currently we can get exactly the same curves right? `faster-coco-eval` vs `modified ultralytics mAP calculation` by @ksv87, so currently the difference might be the calculation we getting the final mAP through these curves. ![image](https://private-user-images.githubusercontent.com/61612323/382743392-fd0dc0ef-541c-42b8-8533-377ff5bc19fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI2OTc2MTgsIm5iZiI6MTczMjY5NzMxOCwicGF0aCI6Ii82MTYxMjMyMy8zODI3NDMzOTItZmQwZGMwZWYtNTQxYy00MmI4LTg1MzMtMzc3ZmY1YmMxOWZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI3VDA4NDgzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkNzdkNDcwYWJkZWJjYWM5NzVmNDYyMGY5OTFmM2UyYWMwNTc2MzYyYjUyMjUyNjdiZGM3ZmQxNzIyZDk2MjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8zsnU8RXsRkXJeINwziqAUVQxWG07nlc1WFYfyk4WbM) In faster-coco-eval, do we get the final mAP by calculating the areas under each curve?

@Laughing-q I noticed that COCOAPI uses `maxDets=100` to calculate the AP, while Ultralytics defaults to 300. COCOAPI selects the first 100 detection scores without sorting and then uses that when there are more than 100 detections for an image.
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.394
```
However, using `max_det=100` during `model.val()`, increases the mAP score for YOLO11n in Ultralytics calculation from 39.2 to 39.5 but lowers the COCOAPI mAP from 39.4 to 39.2.
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.395
```

**Comment by Y-T-G** [comment]
I looked at the COCOAPI code and compared with the Ultralytics evaluation code. There are a few discrepancies besides `max_det` not being equal:

1. Box coordinates and confidence values [are rounded in `predictions.json`](https://github.com/ultralytics/ultralytics/blob/7c97ed19372655a8b3d9050fc0f2ffb2e6a31276/ultralytics/models/yolo/detect/val.py#L293), however they're not rounded in Ultralytics calculation.
2. [`tpc`](https://github.com/ultralytics/ultralytics/blob/7c97ed19372655a8b3d9050fc0f2ffb2e6a31276/ultralytics/utils/metrics.py#L587) (true positive count) is slightly different in COCOAPI vs. Ultralytics. For example, for the person class with `yolo11n.pt`, Ultralytics determines 9520 TPs at 0.50 IOU while COCOAPI determines 9522 TPs at the same IOU.
3. However, the major reason for discrepancy is the [`fpc`](https://github.com/ultralytics/ultralytics/blob/7c97ed19372655a8b3d9050fc0f2ffb2e6a31276/ultralytics/utils/metrics.py#L586) (false positive counts) in MS-COCO. For example, `fpc` in Ultralytics was 68129 while in COCOAPI, it was 59539. This is an issue particularly with MS-COCO dataset. In COCOAPI, [detections that matches with ignored ground truth objects](https://github.com/cocodataset/cocoapi/blob/8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9/PythonAPI/pycocotools/cocoeval.py#L299) such as those labeled as `iscrowd` are not considered false positives ([`dtIg`](https://github.com/cocodataset/cocoapi/blob/8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9/PythonAPI/pycocotools/cocoeval.py#L376)), while they're considered false positives in Ultralytics metrics since Ultralytics discards `iscrowd` tagged ground truths. This results in much higher FP count and hence a lower mAP in Ultralytics calculation.

If [the line in COCOAPI](https://github.com/cocodataset/cocoapi/blob/8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9/PythonAPI/pycocotools/cocoeval.py#L376) is changed to `fps = np.logical_not(tps)` which is essentially what Ultralytics is doing, then the mAP becomes similar to the one by Ultralytics. This is with `max_det=100` and the modification by @ksv87:
```python
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100%|██████████| 5000/5000 [00:00<?, ?
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 313/313 [00:34<00:00,  9
                   all       5000      36335      0.653      0.504      0.542      0.387
```
```python
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.388
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.542
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.422
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.040
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.185
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.400
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.320
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.526
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.566
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.320
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.632
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.763
```

**Comment by Y-T-G** [comment]
@ksv87 Can you open a PR with the fix you used here?

**Comment by ksv87** [comment]
@Y-T-G I did it https://github.com/ultralytics/ultralytics/pull/19738

**Comment by onuralpszr** [comment]
It has been added via this PR https://github.com/ultralytics/ultralytics/pull/24175
